### ISSUE TITLE
Proposal: make count and filter more consistent with each other

### ISSUE THREAD

**Issue by deontologician** [rethinkdb/rethinkdb#4480]
Currently, `count` accepts a function argument:

```coffee
foo.count((x) -> x['bar'])
# equivalent to
foo.filter((x) -> x['bar']).count()
```

But, when you switch to literals, things get ugly:

```coffee
# count how many entries in the stream equal the string "bar"
foo.count("bar")
# count entries in the stream. "bar" is interpreted as true, so nothing is filtered
foo.filter("bar").count()
# count how many objects in the stream have their "bar" field equal to "x"
foo.filter({bar: "x"}).count()
# count how many entries in the stream are equal to the document {bar: "x"}
foo.count({bar: "x"})
```

As discussed in #4459, the behavior for string literals in `.filter` is confusing and probably uncontroversially bad. But I think it would also make sense to bring `.count` into line with filter in accepting pathspecs. While it's a little convenient to be able to do things like `.count(3)`, in practice I bet pathspecs would end up getting used more often.

I was recently bitten by this, I was debugging a `.count` invocation by replacing it with filter to see what was getting weeded out, and `.filter(pathspec).count()` was returning a different number from `.count(pathspec)`

**Comment by mlucy** [comment]
I agree we should make `foo.filter('bar')` not silently turn into `foo`.  I'm not 100% sure whether we should make it error or turn into `foo.filter{|x| x['bar']}`.  I'm also not sure whether `foo.filter(true)` should continue silently turning into `foo`.

I think I would strongly expect `arr.count(3)` to count the number of 3s in `arr`.  That's also the behavior in Python and Ruby.

**Comment by deontologician** [comment]
I think `.filter(3/"foo")` and `.count(3/"foo")` should error. That would correct anyone looking for ruby or python behavior, and unambiguously emphasize that it isn't doing a bracket or anything either. Ruby and Python don't have pathspecs, and in the context of reql I think it's more consistent

**Comment by mlucy** [comment]
Making `r.expr([1, 2, 3]).count(x)` count the number of occurences of `x` seems both intuitive and useful to me.  I would guess most people using Python or Ruby would expect that to work, and having to write `r.expr([1, 2, 3]).count(function(row){ return row.eq(x);})` instead is actually kind of annoying.

I don't think there are a lot of people expecting it to use pathspecs and then being extremely confused when it instead counts literal objects.  (Or, at least, I haven't seen questions about it before.)

Counting the number of occurences also the current behavior, and the bar for removing behavior people might be using is higher than the bar for not including it.

---

I would be moderately strongly opposed to removing the `arr.count(x)` behavior.  @coffeemug, what do you think?

**Comment by coffeemug** [comment]
I agree with @mlucy -- I think `x.count(y)` should count the number of occurrences of `y`; I think it's very useful and the benefits of keeping outweigh the benefits of taking it out IMO.

**Comment by deontologician** [comment]
Well, in ruby it would be `r.expr([1,2,3]).count{|row| row == x}` that example looks bad because it's in js (which would probably use r.row here).

> I don't think there are a lot of people expecting it to use pathspecs and then being extremely confused when it instead counts literal objects. 

Count me among them. It was confusing yesterday when I ran into it, specifically because I was expecting to be able to get it working with filter and then switch it to a count afterward.

I can see where you guys are coming from, `.count(3)` seems intuitive, but `.count("foo")` gets murkier, since it might mean counting objects that have the field foo etc. I feel like reql has developed enough flavor of its own that similarity to python or ruby isn't as good a guideline as consistency with other reql terms.

**Comment by mlucy** [comment]
It looks like there's some genuine disagreement here, so let's delay talking about it until the next ReQL discussion period (which should be pretty soon since the 2.1 beta is so close).

**Comment by marshall007** [comment]
To me, terms like `count` and `contains` seem like the odd-balls here by assuming literals as the first argument. I think `map`, `concat_map`, `order_by`, etc should all consistently accept field-or-function style arguments. Likewise, as @deontologician has suggested in this issue, things like `filter`, `count` and `contains` should accept pathspecs. That being said, perhaps the existing behavior could also be preserved by using `.count(r.literal(3))`?

I'd very much prefer that terms unify around the pathspecs concept rather than accepting literals in specific cases.

As an aside, ReQL has a lot in common with the `lodash` utility library in JS and `lodash` functions are very consistent with regard to the arguments they accept. It might be a good source of inspiration in deciding these issues.

**Comment by mlucy** [comment]
Anecdotally, I write `arr.contains(LITERAL)` at least once a month (often for queries like `r.table('test').filter{|row| row['tags'].contains(SOME_TAG)}`) and have never once wanted to write `arr.contains(PATHSPEC)`.  Checking for set membership seems like a much more common operation than testing whether any objects in an array match a pathspec.

**Comment by deontologician** [comment]
I agree for contains, I use the literal form all the time. Just a note for when we pick this up again, we might consider a literal as a special case of a pathspec, though I have no idea of that's a good generalization
